player/command.c

```diff
@@ -7704,216 +7703,247 @@ void mp_option_change_callback(void *ctx, struct m_config_option *co, int flags,
     if (self_update)
         return;
 
-    if (flags & UPDATE_TERM)
-        mp_update_logging(mpctx, false);
-
-    if (flags & (UPDATE_OSD | UPDATE_SUB_FILT | UPDATE_SUB_HARD)) {
-        for (int n = 0; n < num_ptracks[STREAM_SUB]; n++) {
-            struct track *track = mpctx->current_track[n][STREAM_SUB];
-            struct dec_sub *sub = track ? track->d_sub : NULL;
-            if (sub) {
-                int ret = sub_control(sub, SD_CTRL_UPDATE_OPTS,
-                                      (void *)(uintptr_t)flags);
-                if (ret == CONTROL_OK && flags & (UPDATE_SUB_FILT | UPDATE_SUB_HARD)) {
-                    sub_redecode_cached_packets(sub);
-                    sub_reset(sub);
-                    if (track->selected)
-                        reselect_demux_stream(mpctx, track, true);
+    // Coalesce redundant updates and only keep the newest one.
+    bool drop = false;
+    for (int i = 0; i < mpctx->num_option_callbacks; i++) {
+        if (flags && flags == mpctx->option_callbacks[i].flags)
+            drop = true;
+        if (mpctx->option_callbacks[i].co && opt_ptr == mpctx->option_callbacks[i].co->data)
+            drop = true;
+        if (drop && i < mpctx->num_option_callbacks - 1)
+            mpctx->option_callbacks[i] = mpctx->option_callbacks[i + 1];
+        if (drop && i == mpctx->num_option_callbacks - 1) {
+            mpctx->option_callbacks[i].co = co;
+            mpctx->option_callbacks[i].flags = flags;
+        }
+    }
+
+    if (!drop) {
+        struct mp_option_callback callback = { .co = co, .flags = flags};
+        MP_TARRAY_APPEND(mpctx, mpctx->option_callbacks, mpctx->num_option_callbacks, callback);
+    }
+}
+
+void mp_option_run_callbacks(struct MPContext *mpctx)
+{
+    for (int i = 0; i < mpctx->num_option_callbacks; i++) {
```

Comment:
Whole body is in loop, could we keep to processing in static function, to avoid all this indentaion change?

Reply:
I could put the looping part in playloop directly if that's preferred.

Reply:
function is fine, but all the indent changes are not necessary if we split it into two functions.

Reply:
Could do: `handle_option_callbacks`, loop in there, then call this as `mp_option_run_callbacks(mpctx, index)`. So an extra level of indirection but a nicer diff I suppose.

Reply:
Trust in compiler, it will be inlined.